braindecode/models/atcnet.py

```diff
@@ -0,0 +1,735 @@
+    """Causal 1-dimensional convolution
+
+    Code modified from [1]_ and [2]_.
+
+    Parameters
+    ----------
+    in_channels : int
+        Input channels.
+    out_channels : int
+        Output channels (number of filters).
+    kernel_size : int
+        Kernel size.
+    dilation : int, optional
+        Dilation (number of elements to skip within kernel multiplication).
+        Default to 1.
+    **kwargs :
+        Other keyword arguments to pass to torch.nn.Conv1d, except for
+        `padding`!!
+
+    References
+    ----------
+    .. [1] https://discuss.pytorch.org/t/causal-convolution/3456/4
+    .. [2] https://gist.github.com/paultsw/7a9d6e3ce7b70e9e2c61bc9287addefc
+    """
+    def __init__(
+        self,
+        in_channels,
+        out_channels,
+        kernel_size,
+        dilation=1,
+        **kwargs,
+    ):
+        super().__init__()
+        assert "padding" not in kwargs, (
+            "The padding parameter is controlled internally by "
+            f"{type(self).__name__} class. You should not try to override this"
+            " parameter."
+        )
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.kernel_size = kernel_size
+        self.dilation = dilation
+
+        self.conv = nn.Conv1d(
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=kernel_size,
+            dilation=dilation,
+            padding=(kernel_size - 1) * dilation,
+            **kwargs
+        )
+
+    def forward(self, X):
+        out = self.conv(X)
+        return out[..., :-self.conv.padding[0]]
+
+
+class MaxNormLinear(nn.Module):
+    """Linear layer with MaxNorm constraing on weights.
+
+    Equivalent of Keras tf.keras.Dense(..., kernel_constraint=max_norm())
+    [1, 2]_. Implemented as advised in [3]_.
+
+    Parameters
+    ----------
+    in_features: int
+        Size of each input sample.
+    out_features: int
+        Size of each output sample.
+    bias: bool, optional
+        If set to ``False``, the layer will not learn an additive bias.
+        Default: ``True``.
+
+    References
+    ----------
+    .. [1] https://keras.io/api/layers/core_layers/dense/#dense-class
+    .. [2] https://www.tensorflow.org/api_docs/python/tf/keras/constraints/
+           MaxNorm
+    .. [3] https://discuss.pytorch.org/t/how-to-correctly-implement-in-place-
+           max-norm-constraint/96769
+    """
+    def __init__(
+        self,
+        in_features,
+        out_features,
+        bias=True,
+        max_norm_val=2,
+        eps=1e-5,
+    ):
+        super().__init__()
+        self._max_norm_val = max_norm_val
+        self._eps = eps
+        self.linear = nn.Linear(
+            in_features=in_features,
+            out_features=out_features,
+            bias=bias,
+        )
+
+    def forward(self, X):
+        self._max_norm(self.linear.weight)
+        return self.linear(X)
+
+    def _max_norm(self, w):
```

Comment:
Could you add a small comment here that this will modify w in place? was a bit unsure at first whether it does that

Reply:
Sure! you are right, this should be explicitly said! Thanks for the remark!